src/http/tests/test_login.cpp

```diff
@@ -0,0 +1,300 @@
+	{
+		// `players_online` is a memory table and does not support transactions, so we need to clear it manually
+		// do NOT run this test against a running server's database
+		db.executeQuery("TRUNCATE `players_online`");
+	}
+
+	Database& db = Database::getInstance();
+	DBTransaction transaction;
+
+	std::string_view ip = "74.125.224.72";
+	seconds now = duration_cast<seconds>(system_clock::now().time_since_epoch());
+};
+
+using status = boost::beast::http::status;
+
+BOOST_FIXTURE_TEST_CASE(test_login_missing_email, LoginFixture)
+{
+	auto&& [status, body] = tfs::http::handle_login({{"type", "login"}, {"password", "bar"}}, ip);
+
+	BOOST_TEST(status == status::ok);
+	BOOST_TEST(body.at("errorCode").as_int64() == 3);
+}
+
+BOOST_FIXTURE_TEST_CASE(test_login_account_does_not_exist, LoginFixture)
+{
+	auto&& [status, body] =
+	    tfs::http::handle_login({{"type", "login"}, {"email", "foo@example.com"}, {"password", "bar"}}, ip);
+
+	BOOST_TEST(status == status::ok);
+	BOOST_TEST(body.at("errorCode").as_int64() == 3);
+}
+
+BOOST_FIXTURE_TEST_CASE(test_login_missing_password, LoginFixture)
+{
+	auto&& [status, body] = tfs::http::handle_login({{"type", "login"}, {"email", "foo@example.com"}}, ip);
+
+	BOOST_TEST(status == status::ok);
+	BOOST_TEST(body.at("errorCode").as_int64() == 3);
+}
+
+BOOST_FIXTURE_TEST_CASE(test_login_invalid_password, LoginFixture)
+{
+	BOOST_TEST(db.executeQuery(
+	    "INSERT INTO `accounts` (`name`, `email`, `password`) VALUES ('abc', 'foo@example.com', SHA1('bar'))"));
+
+	auto&& [status, body] =
+	    tfs::http::handle_login({{"type", "login"}, {"email", "foo@example.com"}, {"password", "baz"}}, ip);
+
+	BOOST_TEST(status == status::ok);
+	BOOST_TEST(body.at("errorCode").as_int64() == 3);
+}
+
+BOOST_FIXTURE_TEST_CASE(test_login_missing_token, LoginFixture)
+{
+	BOOST_TEST(db.executeQuery(
+	    "INSERT INTO `accounts` (`name`, `email`, `password`, `secret`) VALUES ('abcd', 'foo@example.com', SHA1('bar'), UNHEX('48656c6c6f21dead'))"));
+
+	auto now = duration_cast<seconds>(system_clock::now().time_since_epoch());
+
+	auto&& [status, body] = tfs::http::handle_login(
+	    {
+	        {"type", "login"},
+	        {"email", "foo@example.com"},
+	        {"password", "bar"},
+	        // token too old, guaranteed to be wrong
+	        {"token", generateToken("", (now.count() / AUTHENTICATOR_PERIOD) - 2)},
+	    },
+	    ip);
+
+	BOOST_TEST(status == status::ok);
+	BOOST_TEST(body.at("errorCode").as_int64() == 6);
+}
+
+BOOST_FIXTURE_TEST_CASE(test_login_success_no_players, LoginFixture)
+{
+	BOOST_TEST(db.executeQuery(
+	    "INSERT INTO `accounts` (`name`, `email`, `password`) VALUES ('defg', 'foo@example.com', SHA1('bar'))"));
+
+	auto&& [status, body] =
+	    tfs::http::handle_login({{"type", "login"}, {"email", "foo@example.com"}, {"password", "bar"}}, ip);
+
+	BOOST_TEST(status == status::ok);
+	auto& characters = body.at("playdata").at("characters").as_array();
+	BOOST_TEST(characters.size() == 0);
+}
```

Comment:
```suggestion
BOOST_FIXTURE_TEST_CASE(test_login_success_no_players, LoginFixture)
{
	BOOST_TEST(db.executeQuery(
	    "INSERT INTO `accounts` (`name`, `email`, `password`) VALUES ('defg', 'foobar@example.com', SHA1('bar'))"));

	auto&& [status, body] =
	    tfs::http::handle_login({{"type", "login"}, {"email", "foobar@example.com"}, {"password", "bar"}}, ip);

	BOOST_TEST(status == status::ok);
	auto& characters = body.at("playdata").at("characters").as_array();
	BOOST_TEST(characters.size() == 0);
}
```

Reply:
need to create a new account here because you already added 3 chars to this test account, thus the check will always fail